src/utils/gcsApi.ts

```diff
@@ -0,0 +1,273 @@
+import * as fs from 'fs';
+import * as path from 'path';
+
+import {
+  Bucket as GCSBucket,
+  Storage as GCSStorage,
+  UploadOptions as GCSUploadOptions,
+} from '@google-cloud/storage';
+import { isDryRun } from 'dryrun';
+
+import { logger as loggerRaw } from '../logger';
+import { reportError, ConfigurationError } from './errors';
+import { checkEnvForPrerequisite, RequiredConfigVar } from './env';
+
+const DEFAULT_MAX_RETRIES = 5;
+const DEFAULT_UPLOAD_METADATA = { cacheControl: `public, max-age=300` };
+
+const IS_DRY_RUN = isDryRun();
+
+const logger = loggerRaw.withScope(`[gcs api]`);
+
+/**
+ * Mapping between file extension regexps and the corresponding content type
+ * that will be set.
+ */
+const CONTENT_TYPES_EXT: Array<[RegExp, string]> = [
+  [/\.js$/, 'application/javascript; charset=utf-8'],
+  [/\.js\.map$/, 'application/json; charset=utf-8'],
+];
+
+/**
+ * Configuration options for the GCS bucket
+ */
+export interface GCSBucketConfig {
+  /** Bucket name */
+  bucketName: string;
+  /** ID of the project containing the bucket   */
+  projectId: string;
+  /** CGS credentials */
+  credentials: { client_email: string; private_key: string };
+  /** Maximum number of retries after unsuccessful request */
+  maxRetries?: number;
+}
+
+/**
+ * Bucket path with associated parameters
+ */
+export interface DestinationPath {
+  /** Path inside the bucket to which files will be uploaded */
+  path: string;
+  /** Metadata to be associated with the files uploaded the path */
+  metadata?: any;
+}
+
+/** Authentication credentials for GCS (pulled from env) */
+interface GCSCreds {
+  /** ID of the GCS project containing the bucket */
+  project_id: string;
+  /** Email address used to identify the service account accessing the bucket */
+  client_email: string;
+  /** API key for service account */
+  private_key: string;
+}
+
+/**
+ * Pulls GCS redentials out of the environment, where they can be stored either
+ * as a path to a JSON file or as a JSON string.
+ *
+ * @returns An object containing the credentials
+ */
+export function getGCSCredsFromEnv(
+  jsonVar: RequiredConfigVar,
+  filepathVar: RequiredConfigVar
+): GCSCreds {
+  // make sure we have at least one of the necessary variables
+  try {
+    checkEnvForPrerequisite(jsonVar, filepathVar);
+  } catch (oO) {
```

Comment:
Tiny thing: everywhere else we use `e` as a var name in error handlers (i.e. `catch (e)...`), let's keep using that for consistency.

Reply:
Okay, though in this case I was specifically trying to indicate, "I'm ignoring this thing." When it's something I'm actually using, I've been calling it `err`, though I can switch to `e` if you'd rather.

(The `oO` is something I picked up from @kamilogorek. 🙂Could switch it to `_` if you think that'd make the "this is getting thrown away" hint clearer.)

Reply:
I love my `o_O` :D

Reply:
@tonyo - Should I make the change to `catch(_)`, or merge as is? Happy to do either.

Reply:
I guess consistency wouldn't hurt here (hence my suggestion about `e` that's used everywhere else), but if you want to explicitly mark it as unused, we can use `_` or `_e`.
`_` may potentially conflict with `lodash` (which is kinda stupid, I think we should replace all of those usages), but I don't have a strong opinion, your call.

Reply:
You should be using `lodash-es` anyway,  as it doesnt have to be treeshaken.

Reply:
> I think we should replace all of those usages

I'll put it on my todo list 😊

Reply:
As for the `catch` variable, I settled for calling it `e` and just leaving a comment about how we're ditching it (and why).